sklearn/datasets/base.py

```diff
@@ -919,3 +920,25 @@ def _fetch_remote(remote, dirname=None):
                       "file may be corrupted.".format(file_path, checksum,
                                                       remote.checksum))
     return file_path
+
+
+def _refresh_cache(path):
+    # REMOVE in v0.23
+    import joblib
+    samples_path = _pkl_filepath(path, "samples")
+    targets_path = _pkl_filepath(path, "targets")
+    msg = "sklearn.externals.joblib is deprecated in 0.21"
+    with warnings.catch_warnings(record=True) as warns:
+        X = joblib.load(samples_path)
+        y = joblib.load(targets_path)
+
+        refresh_needed = any([str(x.message).startswith(msg) for x in warns])
+
+    if refresh_needed:
+        try:
+            joblib.dump(X, samples_path, compress=9)
+            joblib.dump(y, samples_path, compress=9)
+        except IOError:
```

Comment:
when will the IOError happen?

Reply:
when the folder is read-only, as I just mentioned (https://github.com/scikit-learn/scikit-learn/pull/14197#issuecomment-506338854)